File: FlashQuizTests/QuizSessionViewControllerTests.m
//
//  QuizSessionViewControllerTests.m
//  SmallSharpTools
//
//  Created by Brennan Stehling on 7/10/16.
//  Copyright © 2016 FlashQuiz. All rights reserved.
//

#import <XCTest/XCTest.h>

#import "AppDelegate.h"
#import "QuizSessionViewController.h"
#import "StartQuizViewController.h"
#import "Question.h"
#import "QuizSession.h"

CG_INLINE UINavigationController *GetNavigationController() {
    AppDelegate *appDelegate = (AppDelegate *)[UIApplication sharedApplication].delegate;
    return (UINavigationController *)appDelegate.window.rootViewController;
}

@interface QuizSessionViewController (QuizSessionViewControllerTests)

@property (weak, nonatomic) IBOutlet UIButton *submitButton;
@property (weak, nonatomic, nullable) IBOutlet UICollectionView *collectionView;
@property (strong, nonatomic, nonnull) QuizSession *quizSession;

- (IBAction)submitButtonTapped:(id)sender;

- (void)submitAnswer:(NSString *)answer;
- (void)expireAfterTimeout;
- (void)finishQuiz;

@end

@interface QuizSessionViewControllerTests : XCTestCase

@end

@implementation QuizSessionViewControllerTests

- (void)setUp {
    [super setUp];
}

- (void)tearDown {
    [super tearDown];
}

- (void)testRoot {
    UINavigationController *nc = GetNavigationController();
    XCTAssertNotNil(nc);
}

- (void)testViewController1 {
    XCTestExpectation *expectation = [self expectationWithDescription:@"UI"];

    UINavigationController *nc = GetNavigationController();
    XCTAssertNotNil(nc);
    XCTAssertTrue([nc isKindOfClass:[UINavigationController class]]);

    [nc popViewControllerAnimated:NO];
    StartQuizViewController *vc = (StartQuizViewController *)nc.topViewController;
    XCTAssertNotNil(vc);
    XCTAssertTrue([vc isKindOfClass:[StartQuizViewController class]]);

    // Runs the button tap action.
    [vc performSegueWithIdentifier:@"pushQuizSession" sender:vc];

    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0.5 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
        QuizSessionViewController *qs = (QuizSessionViewController *)nc.topViewController;
        XCTAssertNotNil(qs);
        XCTAssertTrue([qs isKindOfClass:[QuizSessionViewController class]]);

        [qs.collectionView reloadData];

        Question *question = qs.quizSession.currentQuestion;

        while (question) {
            [qs submitAnswer:question.answers.firstObject];
            question = qs.quizSession.currentQuestion;
        }

        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0.25 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
            [expectation fulfill];
        });
    });

    NSTimeInterval timeout = 10; // 10 seconds
    [self waitForExpectationsWithTimeout:timeout handler:^(NSError * _Nullable error) {
        // do nothing
    }];
}

- (void)testViewController2 {
    XCTestExpectation *expectation = [self expectationWithDescription:@"UI"];

    UINavigationController *nc = GetNavigationController();
    XCTAssertNotNil(nc);
    XCTAssertTrue([nc isKindOfClass:[UINavigationController class]]);

    [nc popViewControllerAnimated:NO];
    StartQuizViewController *vc = (StartQuizViewController *)nc.topViewController;
    XCTAssertNotNil(vc);
    XCTAssertTrue([vc isKindOfClass:[StartQuizViewController class]]);

    // Runs the button tap action.
    [vc performSegueWithIdentifier:@"pushQuizSession" sender:vc];

    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0.5 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
        QuizSessionViewController *qs = (QuizSessionViewController *)nc.topViewController;
        XCTAssertNotNil(qs);
        XCTAssertTrue([qs isKindOfClass:[QuizSessionViewController class]]);

        [qs.collectionView reloadData];

        Question *question = qs.quizSession.currentQuestion;
        Question *lastQuestion = qs.quizSession.questions.lastObject;

        while (question && ![question isEqual:lastQuestion]) {
            [qs submitAnswer:question.answers.firstObject];
            question = qs.quizSession.currentQuestion;
        }

        [qs expireAfterTimeout];

        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0.25 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
            [expectation fulfill];
        });
    });

    NSTimeInterval timeout = 10; // 10 seconds
    [self waitForExpectationsWithTimeout:timeout handler:^(NSError * _Nullable error) {
        // do nothing
    }];
}

- (void)testViewController3 {
    XCTestExpectation *expectation = [self expectationWithDescription:@"UI"];

    UINavigationController *nc = GetNavigationController();
    XCTAssertNotNil(nc);
    XCTAssertTrue([nc isKindOfClass:[UINavigationController class]]);

    [nc popViewControllerAnimated:NO];
    StartQuizViewController *vc = (StartQuizViewController *)nc.topViewController;
    XCTAssertNotNil(vc);
    XCTAssertTrue([vc isKindOfClass:[StartQuizViewController class]]);

    // Runs the button tap action.
    [vc performSegueWithIdentifier:@"pushQuizSession" sender:vc];

    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0.5 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
        QuizSessionViewController *qs = (QuizSessionViewController *)nc.topViewController;
        XCTAssertNotNil(qs);
        XCTAssertTrue([qs isKindOfClass:[QuizSessionViewController class]]);
        XCTAssertNotNil(qs.collectionView);
        XCTAssertTrue(qs.collectionView.visibleCells.count > 0);

        while (!qs.quizSession.isSessionCompleted) {
            NSIndexPath *indexPath = qs.collectionView.indexPathsForVisibleItems.firstObject;
            [qs.collectionView selectItemAtIndexPath:indexPath animated:NO scrollPosition:UICollectionViewScrollPositionNone];
            [qs submitButtonTapped:qs.submitButton];
        }

        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 0.25 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
            [expectation fulfill];
        });
    });

    NSTimeInterval timeout = 10; // 10 seconds
    [self waitForExpectationsWithTimeout:timeout handler:^(NSError * _Nullable error) {
        // do nothing
    }];
}

@end

File: FlashQuizSwiftTests/QuizSessionViewControllerTests.swift
//
//  QuizSessionViewControllerTests.swift
//  FlashQuiz
//
//  Created by Brennan Stehling on 7/17/16.
//  Copyright © 2016 SmallSharpTools. All rights reserved.
//

import XCTest
import UIKit

class QuizSessionViewControllerTests: XCTestCase {

    override func setUp() {
        super.setUp()
    }

    override func tearDown() {
        super.tearDown()
    }

    func testRoot() {
        let nc = getNavigationController()
        XCTAssertNotNil(nc)
    }

    func testViewController1() {
        let expectation = self.expectation(description: "UI")

        guard let nc = getNavigationController() else {
            XCTFail("Failed to get navigation controller")
            return
        }

        nc.popToRootViewController(animated: false)

        guard let vc = nc.topViewController as? StartQuizViewController else {
            XCTFail()
            return
        }

        vc.performSegue(withIdentifier: "pushQuizSession", sender: vc)

        DispatchQueue.main.asyncAfter(deadline: whenInSeconds(1.5)) {
            guard let qs = nc.topViewController as? QuizSessionViewController,
                let collectionView = qs.collectionView else {
                    XCTFail()
                    return
            }

            collectionView.reloadData()

            var currentQuestion = qs.quizSession.currentQuestion
            while currentQuestion != nil {
                guard let question = currentQuestion,
                    let answer = question.answers.first else {
                        XCTFail()
                        break
                }
                qs.submitAnswer(answer)
                currentQuestion = qs.quizSession.currentQuestion
            }

            DispatchQueue.main.asyncAfter(deadline: self.whenInSeconds(0.75)) {
                expectation.fulfill()
            }
        }

        let timeout: TimeInterval = 10
        self.waitForExpectations(timeout: timeout) { (error) in
            // do nothing
        }
    }

    func testViewController2() {
        let expectation = self.expectation(description: "UI")

        guard let nc = getNavigationController() else {
            XCTFail("Failed to get navigation controller")
            return
        }

        nc.popToRootViewController(animated: false)

        guard let vc = nc.topViewController as? StartQuizViewController else {
            XCTFail()
            return
        }

        vc.performSegue(withIdentifier: "pushQuizSession", sender: vc)

        DispatchQueue.main.asyncAfter(deadline: whenInSeconds(1.5)) {
            guard let qs = nc.topViewController as? QuizSessionViewController,
                let collectionView = qs.collectionView else {
                    XCTFail()
                    return
            }

            collectionView.reloadData()

            var currentQuestion = qs.quizSession.currentQuestion
            guard let lastQuestion = qs.quizSession.questions.last else {
                XCTFail()
                return
            }

            // submit all answers except the last one
            while currentQuestion != nil && currentQuestion != lastQuestion {
                guard let question = currentQuestion,
                    let answer = question.answers.first else {
                        XCTFail()
                        break
                }
                qs.submitAnswer(answer)
                currentQuestion = qs.quizSession.currentQuestion
            }

            qs.expireAfterTimeout()

            DispatchQueue.main.asyncAfter(deadline: self.whenInSeconds(0.75)) {
                expectation.fulfill()
            }
        }

        let timeout: TimeInterval = 10
        self.waitForExpectations(timeout: timeout) { (error) in
            // do nothing
        }
    }

    func testViewController3() {
        let expectation = self.expectation(description: "UI")

        guard let nc = getNavigationController() else {
            XCTFail("Failed to get navigation controller")
            return
        }

        nc.popToRootViewController(animated: false)

        guard let vc = nc.topViewController as? StartQuizViewController else {
            XCTFail()
            return
        }

        vc.performSegue(withIdentifier: "pushQuizSession", sender: vc)

        DispatchQueue.main.asyncAfter(deadline: whenInSeconds(1.5)) {
            guard let qs = nc.topViewController as? QuizSessionViewController,
                let collectionView = qs.collectionView else {
                    XCTFail()
                    return
            }

            while !qs.quizSession.isSessionCompleted {
                guard let indexPath = collectionView.indexPathsForVisibleItems.first else {
                        XCTFail()
                        break
                }
                collectionView.selectItem(at: indexPath, animated: false, scrollPosition: UICollectionViewScrollPosition())
                qs.submitButtonTapped(qs.submitButton)
            }

            DispatchQueue.main.asyncAfter(deadline: self.whenInSeconds(0.75)) {
                expectation.fulfill()
            }
        }

        let timeout: TimeInterval = 10
        self.waitForExpectations(timeout: timeout) { (error) in
            // do nothing
        }
    }

    // MARK: Private -

    fileprivate func whenInSeconds(_ seconds: Double) -> DispatchTime {
        return DispatchTime.now() + Double(Int64(seconds * Double(NSEC_PER_SEC))) / Double(NSEC_PER_SEC)
    }

    fileprivate func getNavigationController() -> UINavigationController? {
        if let appDelegate = UIApplication.shared.delegate as? AppDelegate,
        let navigationController = appDelegate.window.rootViewController as? UINavigationController {
            return navigationController
        }

        return nil
    }

}
